public/app/features/dashboard/dashgrid/PanelPerformanceMonitor.tsx

```diff
@@ -0,0 +1,159 @@
+    savedFieldConfig.current = props.panelFieldConfig;
+  }, [props.panelFieldConfig]);
+
+  const logEvent = (eventName: string, newKey: string, newVal: string, oldVal?: string) => {
+    const logObj = {
+      key: newKey,
+      newValue: newVal,
+      oldValue: oldVal ?? '',
+      panelTitle: props.panelTitle,
+      panelId: panelIdStr,
+      panelType: props.panelType,
+    };
+
+    //TODO: do batch push when that becomes available
+    faro.api.pushEvent(eventName, logObj);
+  };
+
+  const logPanelOptionChanges = (panelOptions: any, oldPanelOptions: any) => {
+    for (const [key, value] of Object.entries(panelOptions)) {
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string = typeof value !== 'string' ? JSON.stringify(oldPanelOptions[key]) : oldPanelOptions[key];
+
+      if (oldPanelOptions[key] === undefined) {
+        logEvent('new panel option', key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent('panel option changed', key, newValue, oldValue);
+      }
+    }
+  };
+
+  const logFieldConfigChanges = (fieldConfig: FieldConfigSource, oldFieldConfig: FieldConfigSource) => {
+    // overrides are an array of objects, so stringify it all and log changes
+    // in lack of an index, we can't tell which override changed
+    if (oldFieldConfig.overrides !== fieldConfig.overrides) {
+      logEvent(
+        'field config overrides changed',
+        'overrides',
+        JSON.stringify(fieldConfig.overrides),
+        JSON.stringify(oldFieldConfig.overrides)
+      );
+    }
+
+    const oldDefaults: { [key: string]: any } = oldFieldConfig.defaults;
+
+    // go through field config keys except custom, we treat that below
+    for (const [key, value] of Object.entries(fieldConfig.defaults)) {
+      if (key === 'custom') {
+        continue;
+      }
+
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string = typeof value !== 'string' ? JSON.stringify(oldDefaults[key]) : oldDefaults[key];
+
+      if (oldDefaults[key] === undefined) {
+        logEvent('new default field config', key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent('default field config changed', key, newValue, oldValue);
+      }
+    }
+
+    if (!fieldConfig.defaults.custom) {
+      return;
+    }
+
+    // go through custom field config keys
+    for (const [key, value] of Object.entries(fieldConfig.defaults.custom)) {
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string =
+        typeof value !== 'string' ? JSON.stringify(oldDefaults.custom[key]) : oldDefaults.custom[key];
+
+      if (oldDefaults.custom[key] === undefined) {
+        logEvent('new custom field config', key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent('custom field config changed', key, newValue, oldValue);
```

Comment:
we may want to consider consolidating event types as consts for better maintainability (maybe grouped in a separate file that we import them from?)